rllib/utils/actor_manager.py

```diff
@@ -0,0 +1,324 @@
+    def __fetch_result_and_mark_state(
+        self,
+        *,
+        remote_actor_indices: List[int],
+        remote_calls: List[ray.ObjectRef],
+        timeout_seconds: int = None,
+    ) -> Tuple[List[ray.ObjectRef], List[T]]:
+        """Try fetching results from remote actor calls.
+
+        Mark whether a actor is healthy or not accordingly.
+
+        Args:
+            remote_actor_indices: indices of the actors these remote
+                calls were fired against.
+            remote_calls: list of remote calls to fetch.
+            timeout_seconds: timeout for the ray.wait() call. Default is None.
+
+        Returns:
+            A Tuple of:
+                The remote object refs that are ready within timeout.
+                Results for these remote calls.
+            Note that the list of object refs is alwasy the same length as the
+            list of results.
+            But some of the results may be Errors.
+        """
+        # Notice that we do not return the refs to any unfinished calls to the
+        # user, since it is not safe to handle such remote actor calls outside the
+        # context of this actor manager. These requests are simply dropped.
+        timeout = float(timeout_seconds) if timeout_seconds is not None else None
+        ready, _ = ray.wait(
+            remote_calls,
+            num_returns=len(remote_calls),
+            timeout=timeout,
+            # Make sure remote results are fetched locally in parallel.
+            fetch_local=True,
+        )
+
+        if timeout and len(ready) < len(remote_calls):
+            logger.error(
+                f"Failed to wait for {len(remote_calls) - len(ready)} remote calls."
+            )
+
+        # Remote data should already be fetched to local object store at this point.
+        remote_results = []
+        for i, r in enumerate(ready):
+            actor_idx = remote_actor_indices[i]
+            try:
+                remote_results.append(ray.get(r))
+                # Able to fetch return value. Mark this actor active if necessary.
+                if not self.is_actor_healthy(actor_idx):
+                    print(f"brining actor {actor_idx} back into service.")
+                self.set_actor_state(actor_idx, healthy=True)
+            except RayActorError as e:
+                remote_results.append(e)
+                # Take this actor out of service and wait for Ray Core to recover it.
+                if self.is_actor_healthy(actor_idx):
+                    print(
+                        f"Ray error, taking actor {actor_idx} out of service. {str(e)}"
+                    )
+                self.set_actor_state(actor_idx, healthy=False)
+            except Exception as e:
+                # Application error. Return Exception as-is to the caller.
+                remote_results.append(e)
+
+        return ready, remote_results
+
+    @DeveloperAPI
+    def foreach_actor(
+        self,
+        func: Callable[[T], T],
```

Comment:
What does this type `Callable[[T], T]` actually mean? `func(foo: int) -> int` is allowed but `func(foo: int) -> float` is not?